test: update succinct tests to follow our conventions

stevennevins opened the pull request:
**Description**

Port tests for OPSuccint game and adapt to patterns in the monorepo

**Tests**

Port over test coverage for OPSuccinct and add test coverage


**Metadata**

Closes https://github.com/ethereum-optimism/optimism/issues/18461

stevennevins committed:
test: update succinct tests to follow our conventions

stevennevins committed:
fix: test naming and remove unused imports

stevennevins committed:
chore: add AccessManager snapshots

Contract changed from abstract to concrete, now needs snapshot files.

stevennevins committed:
fix: specify GameOver revert in expectRevert

stevennevins committed:
fix: snapshot for access manager

stevennevins committed:
fix: create interface for contract

stevennevins committed:
fix: snapshot semver bump

stevennevins committed:
fix: bump semver

stevennevins committed:
chore: bump semver

stevennevins committed:
fix: semver lock mismatch

stevennevins committed:
fix: enforce strict pragma for deterministic init code hashes

stevennevins committed:
chore: fix pragma

stevennevins committed:
fix: lock pragma versions

stevennevins committed:
fix: semver lock passing

stevennevins committed:
chore: cleanup whitespace

falcorocks commented:
hi @stevennevins , sorry for the bake errors. I've fixed it in #18498, if you rebase your branch you should be good!

stevennevins commented:
> hi @stevennevins , sorry for the bake errors. I've fixed it in #18498, if you rebase your branch you should be good!

No worries! I was trying to resolve separate issues with CI. Will rebase  those changes today tho

Inphi commented:
I think you only need strict versions pragmas for actual contracts. Libraries and abstract contracts, since they do not produce concrete abi artifacts, can have a fuzzy solc version specified. The actual contract that imports them will use a strict solc version.
So the diff here can be much smaller.

stevennevins committed:
chore: bump semver

stevennevins committed:
chore: reduce diff

stevennevins committed:
fix: strict pragma

stevennevins commented:
> I think you only need strict versions pragmas for actual contracts. Libraries and abstract contracts, since they do not produce concrete abi artifacts, can have a fuzzy solc version specified. The actual contract that imports them will use a strict solc version. So the diff here can be much smaller.

Yeah the diff kind of spiraled when it was causing failures in the semver-lock.json and some obscure issues with vm.getCode

stevennevins committed:
feat: fix for ci and compiler profile

stevennevins committed:
chore: forge fmt

stevennevins committed:
fix: regenerate semver-lock with clean build

The semver-lock.json had incorrect hashes due to stale build
artifacts. Clean build restores hashes to match develop.

stevennevins committed:
fix: semgrep for proxy

stevennevins committed:
fix: workaround for potential cache issue

stevennevins committed:
chore: clean build semver bump

stevennevins committed:
fix: remove troubleshooting

stevennevins committed:
chore: bump semver

stevennevins committed:
fix: compiler setting issues

stevennevins committed:
chore: remove helper for vm.getCode

stevennevins commented on packages/contracts-bedrock/src/dispute/zk/AccessManager.sol:
```diff
@@ -9,7 +9,7 @@ import { OP_SUCCINCT_FAULT_DISPUTE_GAME_TYPE } from "src/dispute/lib/Types.sol";
 /// @title AccessManager
 /// @notice Manages permissions for dispute game proposers and challengers.
-abstract contract AccessManager is Ownable {
+contract AccessManager is Ownable {
```
we'll be removing AccessManager in a follow up. resolving

Inphi commented on packages/contracts-bedrock/foundry.toml:
```diff
@@ -25,7 +25,6 @@ compilation_restrictions = [
   { paths = "src/dispute/v2/FaultDisputeGameV2.sol", optimizer_runs = 5000 },
   { paths = "src/dispute/PermissionedDisputeGame.sol", optimizer_runs = 5000 },
   { paths = "src/dispute/v2/PermissionedDisputeGameV2.sol", optimizer_runs = 5000 },
-  { paths = "src/dispute/zk/OPSuccinctFaultDisputeGame.sol", optimizer_runs = 5000 },
```
I'm still not sure why this change is necessary. When I revert this locally, the build and tests work fine.

stevennevins replied:
were you running all the tests? it causes an init code hash mismatch if you revert this

it was happneing when i run the just pre-pr for the semver lock checks and if i run all the tests

Inphi replied:
This is OK for now. I'll create an issue to figure out the problem here later.

Inphi commented on packages/contracts-bedrock/test/dispute/zk/OPSuccinctFaultDisputeGame.t.sol:
```diff
@@ -0,0 +1,787 @@
+    GameNotOver,
+    IncorrectDisputeGameFactory
+} from "src/dispute/lib/Errors.sol";
+import { OP_SUCCINCT_FAULT_DISPUTE_GAME_TYPE } from "src/dispute/lib/Types.sol";
```
nit: We can follow up in a later PR. The `OP_SUCCINCT_FAULT_DISPUTE_GAME_TYPE` should be renamed to `OPTIMISTIC_ZK` to match the challenger's.

stevennevins replied:
Looks like also op-challenger is using game type 10? will add that change as well

Inphi commented on packages/contracts-bedrock/test/dispute/zk/OPSuccinctFaultDisputeGame.t.sol:
```diff
@@ -0,0 +1,787 @@
+
+/// @title OPSuccinctFaultDisputeGame_TestInit
+/// @notice Base test contract with shared setup for OPSuccinctFaultDisputeGame tests.
+abstract contract OPSuccinctFaultDisputeGame_TestInit is Test {
```
The test setup in this contract is doing too much work that is already handled by `CommonTest` to setup various OP contracts. It would be cleaner for this contract to reuse the test facility provided by [`DIsputeGameFactory_TestInit`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/dispute/DisputeGameFactory.t.sol#L51) which uses `CommonTest`.
You can then add a helper function similar to [`setupFaultDisputeGameV1`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/dispute/DisputeGameFactory.t.sol#L198) to the `DisputeGameFactory_TestInit` contract so that you can add the optimistic-zk game implementation to the factory for this test.

stevennevins committed:
refactor: reuse DisputeGameFactory_TestInit in OPSuccinct tests

Remove duplicated setup logic from OPSuccinctFaultDisputeGame tests
by inheriting from DisputeGameFactory_TestInit instead of Test.

- Add setupOPSuccinctFaultDisputeGame() helper to DisputeGameFactory_TestInit
- Add OPSuccinctGameParams struct for configurable game parameters
- Remove ~100 lines of duplicated contract deployment code
- Update variable references to use inherited contracts

stevennevins committed:
refactor: rename OPSuccinctFaultDisputeGame to OptimisticZkGame

Align contract naming with op-challenger's OptimisticZKGameType.

Renames:
- IOPSuccinctFaultDisputeGame -> IOptimisticZkGame
- OPSuccinctFaultDisputeGame -> OptimisticZkGame
- Test contracts and helper functions updated accordingly

Inphi commented on packages/contracts-bedrock/src/dispute/zk/OptimisticZkGame.sol:
```diff
@@ -43,10 +43,10 @@ import { IAnchorStateRegistry } from "interfaces/dispute/IAnchorStateRegistry.so
 /// @notice An implementation of the `IFaultDisputeGame` interface.
 /// @dev Derived from https://github.com/succinctlabs/op-succinct (at commit c13844a9bbc330cca69eef2538d8f8ec123e1653)
-contract OPSuccinctFaultDisputeGame is Clone, ISemver, IDisputeGame {
+contract OptimisticZkGame is Clone, ISemver, IDisputeGame {
```
nit: For consistency

```suggestion
contract OptimisticZkDisputeGame is Clone, ISemver, IDisputeGame {
```

stevennevins committed:
chore: remove unused import

stevennevins committed:
chore: bump snapshots

Inphi reviewed (approved):
Looks good overall.

stevennevins committed:
fix: update abi_loader.go embed path after rename

stevennevins committed:
fix: update snapshots

stevennevins committed:
fix: circle ci cache fix

stevennevins committed:
fix: ci cache

ajsutton commented on packages/contracts-bedrock/src/dispute/lib/Types.sol:
```diff
@@ -127,10 +127,10 @@ library LocalPreimageKey {
+//               `OptimisticZkGame` Types                   //
 ////////////////////////////////////////////////////////////////
 
-uint32 constant OP_SUCCINCT_FAULT_DISPUTE_GAME_TYPE = 42;
```
I think we want to keep the `OP_SUCCINCT_FAULT_DISPUTE_GAME_TYPE` as is and add a new `OPTIMISTIC_ZK_GAME_TYPE = 10`  back up after line 80.

This constant is basically a place holder for actual op-succinct deployments and we're adding a new game type when we bring it into the monorepo and adjust it to be a generic ZK game type.

stevennevins replied:
sgtm will add back

stevennevins committed:
chore: add back placeholder for opsuccinet

stevennevins committed:
chore: remove cache-lite

stevennevins committed:
fix: make OptimisticZkGame tests compatible with fork mode

Replace hardcoded block numbers and game indices with dynamically
calculated values based on anchor state. This allows tests to pass
in both regular mode (empty factory) and fork mode (mainnet fork
with existing games).

stevennevins committed:
fix: skip OptimisticZkGame tests on fork

Tests create games with specific state assumptions that conflict
with existing state on forked networks.

stevennevins committed:
fix: use same pattern as fault dispute game

stevennevins committed:
chore: check CI skpping for tests for op zk

ajsutton commented on packages/contracts-bedrock/src/dispute/lib/Types.sol:
```diff
@@ -127,9 +127,10 @@ library LocalPreimageKey {
+//               `OptimisticZkGame` Types                   //
 ////////////////////////////////////////////////////////////////
 
+uint32 constant OPTIMISTIC_ZK_GAME_TYPE = 10;
```
Why isn't this:
```
GameType internal constant OPTIMISTIC_ZK = GameType.wrap(10);
```

like the other game types back up on line 80?  I don't really know why the OP_SUCCINCT 42 isn't like that either but seems confusing to have some types defined as a `GameType` and some as `uint32`.

stevennevins replied:
yeah good call. will align

ajsutton reviewed (approved):
LGTM. Just the one nit.

stevennevins committed:
fix: remove OPSuccinctFaultDisputeGame from compiler restrictions

stevennevins committed:
fix: move OptimisticZk type into GameTypes and update usage

stevennevins committed:
fix: bump semver
